cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,495 @@
+    id_segments.push_back(graph_view.get_local_adj_matrix_partition_row_last(i));
+    // Count of relative position of the vertices
+    vertex_count_offsets.push_back(counter);
+    // Adjacency list offset pointer of each partition
+    adjacency_list_offsets.push_back(matrix_partition.get_offsets());
+    // Adjacency list indices pointer of each partition
+    adjacency_list_indices.push_back(matrix_partition.get_indices());
+
+    counter += matrix_partition.get_major_size();
+  }
+
+  // Allocate device memory for transfer
+  rmm::device_uvector<vertex_t> r_segments(id_segments.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> vc_offsets(vertex_count_offsets.size(), handle.get_stream());
+  rmm::device_uvector<edge_t const*> al_offsets(adjacency_list_offsets.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t const*> al_indices(adjacency_list_indices.size(),
+                                                  handle.get_stream());
+
+  // Transfer data
+  raft::update_device(
+    r_segments.data(), id_segments.data(), id_segments.size(), handle.get_stream());
+  raft::update_device(vc_offsets.data(),
+                      vertex_count_offsets.data(),
+                      vertex_count_offsets.size(),
+                      handle.get_stream());
+  raft::update_device(al_offsets.data(),
+                      adjacency_list_offsets.data(),
+                      adjacency_list_offsets.size(),
+                      handle.get_stream());
+  raft::update_device(al_indices.data(),
+                      adjacency_list_indices.data(),
+                      adjacency_list_indices.size(),
+                      handle.get_stream());
+
+  return std::make_tuple(
+    std::move(r_segments), std::move(vc_offsets), std::move(al_offsets), std::move(al_indices));
+}
+
+}  // namespace detail
+
+/**
+ * @brief Return global out degrees of active sources
+ *
+ * Get partition information of all graph partitions on the gpu and select
+ * global degrees of all active sources
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param active_sources Device vector containing all the vertex id that are processed by
+ * gpus in the column communicator
+ * @param global_out_degrees Global out degrees for every source represented by current gpu
+ * @return Global out degrees of all sources in active_sources
+ */
+template <typename GraphViewType>
+rmm::device_uvector<typename GraphViewType::edge_type> get_active_source_global_degrees(
+  raft::handle_t const& handle,
+  GraphViewType const& graph_view,
+  rmm::device_uvector<typename GraphViewType::vertex_type>& active_sources,
+  const rmm::device_uvector<typename GraphViewType::edge_type>& global_out_degrees)
```

Comment:
And should we reveal the detail of edge partitioning to end-users through our public API? We may just let users think about vertex partitioning and create a second version of compute_out_degrees
(https://github.com/rapidsai/cugraph/blob/branch-22.04/cpp/include/cugraph/graph_view.hpp#L591) taking a vertex list.

Creating a source global out degrees from local vertex's global out degrees should be handled in the detail space (or inside the implementation of public API).

Reply:
And if we move this function to the detail space, source here should be renamed to major.